zebra-state/src/service/finalized_state/merge_operator.rs

```diff
@@ -0,0 +1,51 @@
+//! Merge operators for Zebra's RocksDB column families.
+//!
+//! These functions restrict the operations that are allowed on a RockDB column family.
+//! They are used as callbacks during RocksDB transactions and database writes.
+//!
+//! For design details, see:
+//! https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#rocksdb-data-structures
+//!
+//! And for the API documentation, see:
+//! https://github.com/facebook/rocksdb/wiki/Merge-Operator#what
+//! https://docs.rs/rocksdb/*/rocksdb/merge_operator/index.html
+
+use rocksdb::merge_operator::MergeOperands;
+
+/// Reject updates after the first value is inserted for each key.
+///
+/// If an `existing_val` for `new_key` is already present in the column family,
+/// or there are multiple writes to `new_key`, reject this update.
+///
+/// This function is trivially associative, because it keeps the first value for each key.
+///
+/// Note: this function does not change how deletes are handled.
+///
+/// # Safety
+///
+/// This function is called from C, so it should avoid panicking.
+/// If it panics, it must abort rather than unwinding.
+pub(super) fn reject_updates_to_existing_values(
+    _key: &[u8],
+    existing_val: Option<&[u8]>,
+    operands: &mut MergeOperands,
+) -> Option<Vec<u8>> {
+    // reject overwrites to keys that already have values
+    if existing_val.is_some() {
+        return None;
+    }
+
+    if operands.size_hint().0 > 1 {
```

Comment:
Warning: Using `size_hint` might be a little dangerous. This will currently work because the bounds are exact in [the source code](https://docs.rs/rocksdb/0.16.0/src/rocksdb/merge_operator.rs.html#216), and I think there's no reason to expect this to change in the future.

However, if it does change in the future, the `== 0` and the `== 1` cases might be incorrect. For example, the bounds could be `0..` (`(0, None)`) or `1..3` (`(1, Some(3))`).  It might be more cautious to do something like:

```
let value_to_insert = operands.next()?;

if operands.next().is_some() {
    // reject partial merges and combined writes of multiple keys
    None
} else {
    // if there is a single value, write it
    Some(value_to_insert)
}
```

Reply:
Yes I noticed, it should really implement `ExactSizeIterator`.

In the meantime you're right, we should code defensively.

Reply:
eccaaeb7

Reply:
This change is obsolete, because the initial design didn't work.